### PR TITLE
A MethodHandles API privateLookupIn

### LINKED ISSUE #41: OpenJ9 does not work with IntellijIDEA OSS from jetbrains
When I try to launch IntellijIDEA with the OpenJ9 JRE, I get the following error:

```
Internal error. Please report to http://jb.gg/ide/critical-startup-errors

io.netty.channel.ChannelException: Unable to create Channel from class class io.netty.channel.socket.nio.NioServerSocketChannel
    at io.netty.channel.ReflectiveChannelFactory.newChannel(ReflectiveChannelFactory.java:40)
    at io.netty.bootstrap.AbstractBootstrap.initAndRegister(AbstractBootstrap.java:321)
    at io.netty.bootstrap.AbstractBootstrap.doBind(AbstractBootstrap.java:283)
    at io.netty.bootstrap.AbstractBootstrap.bind(AbstractBootstrap.java:279)
    at io.netty.bootstrap.AbstractBootstrap.bind(AbstractBootstrap.java:268)
    at org.jetbrains.io.BuiltInServer.bind(BuiltInServer.java:149)
    at org.jetbrains.io.BuiltInServer.start(BuiltInServer.java:118)
    at org.jetbrains.io.BuiltInServer.startNioOrOio(BuiltInServer.java:105)
    at com.intellij.idea.SocketLock.lambda$lock$2(SocketLock.java:153)
    at com.intellij.idea.SocketLock$$Lambda$189.000000007C182CE0.call(Unknown Source)
    at com.intellij.idea.SocketLock.underLocks(SocketLock.java:181)
    at com.intellij.idea.SocketLock.lock(SocketLock.java:128)
    at com.intellij.idea.StartupUtil.lockSystemFolders(StartupUtil.java:286)
    at com.intellij.idea.StartupUtil.prepareAndStart(StartupUtil.java:122)
    at com.intellij.idea.MainImpl.start(MainImpl.java:34)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(java.base@9-internal/Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(java.base@9-internal/NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@9-internal/DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(java.base@9-internal/Method.java:564)
    at com.intellij.ide.plugins.PluginManager.lambda$start$0(PluginManager.java:95)
    at com.intellij.ide.plugins.PluginManager$$Lambda$179.0000000024E69BD0.run(Unknown Source)
    at java.lang.Thread.run(java.base@9-internal/Thread.java:835)
Caused by: java.lang.NoClassDefFoundError: io.netty.buffer.ByteBufAllocator (initialization failure)
    at java.lang.J9VMInternals.initializationAlreadyFailed(java.base@9-internal/J9VMInternals.java:141)
    at io.netty.channel.DefaultChannelConfig.<init>(DefaultChannelConfig.java:56)
    at io.netty.channel.DefaultChannelConfig.<init>(DefaultChannelConfig.java:69)
    at io.netty.channel.socket.DefaultServerSocketChannelConfig.<init>(DefaultServerSocketChannelConfig.java:48)
    at io.netty.channel.socket.nio.NioServerSocketChannel$NioServerSocketChannelConfig.<init>(NioServerSocketChannel.java:195)
    at io.netty.channel.socket.nio.NioServerSocketChannel$NioServerSocketChannelConfig.<init>(NioServerSocketChannel.java:193)
    at io.netty.channel.socket.nio.NioServerSocketChannel.<init>(NioServerSocketChannel.java:87)
    at io.netty.channel.socket.nio.NioServerSocketChannel.<init>(NioServerSocketChannel.java:72)
    at java.lang.J9VMInternals.newInstanceImpl(java.base@9-internal/Native Method)
    at java.lang.Class.newInstance(java.base@9-internal/Class.java:1954)
    at io.netty.channel.ReflectiveChannelFactory.newChannel(ReflectiveChannelFactory.java:38)
    ... 21 more
Caused by: java.lang.UnsupportedOperationException: The method has not yet been implemented for now
    at java.lang.invoke.MethodHandles.privateLookupIn(java.base@9-internal/MethodHandles.java:1538)
    at java.util.concurrent.atomic.Striped64$1.run(java.base@9-internal/Striped64.java:390)
    at java.util.concurrent.atomic.Striped64$1.run(java.base@9-internal/Striped64.java:387)
    at java.security.AccessController.doPrivileged(java.base@9-internal/AccessController.java:638)
    at java.util.concurrent.atomic.Striped64.<clinit>(java.base@9-internal/Striped64.java:386)
    at io.netty.util.internal.PlatformDependent.newLongCounter(PlatformDependent.java:318)
    at io.netty.buffer.PoolArena.<init>(PoolArena.java:67)
    at io.netty.buffer.PoolArena$HeapArena.<init>(PoolArena.java:671)
    at io.netty.buffer.PooledByteBufAllocator.<init>(PooledByteBufAllocator.java:236)
    at io.netty.buffer.PooledByteBufAllocator.<init>(PooledByteBufAllocator.java:187)
    at io.netty.buffer.PooledByteBufAllocator.<init>(PooledByteBufAllocator.java:176)
    at io.netty.buffer.PooledByteBufAllocator.<init>(PooledByteBufAllocator.java:162)
    at io.netty.buffer.PooledByteBufAllocator.<clinit>(PooledByteBufAllocator.java:143)
    at io.netty.buffer.ByteBufUtil.<clinit>(ByteBufUtil.java:79)
    at io.netty.buffer.ByteBufAllocator.<clinit>(ByteBufAllocator.java:24)
    ... 31 more
```

IntellijIDEA can be downloaded from [here](https://www.jetbrains.com/idea/download/index.html#section=linux). I am using the community edition. My OpenJ9 --version is
```
openjdk 9-internal
OpenJDK Runtime Environment (build 9-internal+0-adhoc.jenkins.openjdk)
Eclipse OpenJ9 VM (build 2.9, JRE 9 Linux amd64-64 Compressed References 20170915_6 (JIT enabled, AOT enabled)
J9VM - cea1ed7
JIT  - cea1ed7
OMR  - 617de12
OpenJDK  - 83f5cd0 based on )
```

### PR DESCRIPTION
**Pull request by ChengJin01**
The code is to introduce a new API called
privateLookupIn to produce a lookup object
with private access to the  request class.

fixes #41

Signed-off-by: CHENGJin <jincheng@ca.ibm.com>

### CODE REVIEW

**Comment by mstoodle** [comment]
Please include a reference to the issue #41 in the pull request text and commit message, just to make the connection clear. And if this pull request will resolve #41, then please add the Fixes before the reference.

**Comment by ChengJin01** [comment]
The code can't reviewed now as it depends on the changes the MODULE bits which will be created via another PR (needs to get merged first before this PR).

**Comment by genie-openj9** [comment]
Can one of the admins verify this patch?

**Comment by genie-openj9** [comment]
Can one of the admins verify this patch?

**Comment by genie-openj9** [comment]
Can one of the admins verify this patch?

**Comment by ChengJin01** [comment]
The changes for adding the MODULE bit has been created at https://github.com/eclipse/openj9/pull/81 for review. They must get merged first before the current changes of privateLookupIn (still need to updated with test cases in the new test project after that)

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 1]
```diff
@@ -1667,20 +1671,67 @@ static MethodHandle filterArgument(MethodHandle target, int pos, MethodHandle fi
...
-	 * @throws NullPointerException - if any of the arguments are null
+	 * @throws NullPointerException - if any of the arguments is null
 	 * @throws IllegalAccessException - if access checking fails
 	 * @throws SecurityException - if the SecurityManager prevents access
 	 */
 	public static MethodHandles.Lookup privateLookupIn(Class<?> targetClass, MethodHandles.Lookup callerLookup) throws IllegalArgumentException, NullPointerException, IllegalAccessException, SecurityException {
-		throw new UnsupportedOperationException("The method has not yet been implemented for now"); //$NON-NLS-1$
+		if ((null == targetClass) || (null == callerLookup)) {
+			/*[MSG "K065S", "Both the requested class and the caller lookup must be non-null"]*/
+			System.out.println("NullPointerException K065S: Both the requested class and the caller lookup must be non-null"); //$NON-NLS-1$
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Remove the System.out.println call.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 2]
```diff
@@ -1667,20 +1671,67 @@ static MethodHandle filterArgument(MethodHandle target, int pos, MethodHandle fi
...
+			throw new NullPointerException(com.ibm.oti.util.Msg.getString("K065S")); //$NON-NLS-1$
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: This is a good candidate for `Objects.notNull()`

**Comment by ChengJin01** [comment]
The code here needs to be double-checked and clean up all unnecessary printing messages after adding new test cases before code review.

**Comment by ChengJin01** [comment]
The code/test cases were verified via personal build and now are ready for review.

test framework reviewer:  @renfeiw,
peer reviewer: @JasonFengJ9 
senior reviewer: @DanHeidinga

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 1 above)
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: These printing messages were used for test and already removed.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 2 above)
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Already replaced with Objects.isNull()

File: test/OpenJ9_Jsr_292_API/playlist.xml [hunk 3]
```diff
@@ -38,12 +39,13 @@
...
 	
-		<test>
-		<testCaseName>openj9_jsr292Test_JitCount0_SE90</testCaseName>
-		<variations>
+	<test>
+	<testCaseName>openj9_jsr292Test_JitCount0_SE90</testCaseName>
+	<variations>
 			<variation>-Xjit:count=0</variation>
-		</variations>
+	</variations>
```
Comment (renfeiw) [inline test/OpenJ9_Jsr_292_API/playlist.xml]: Please add one more indentation before <testCaseName> <variations> and </variations>.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 4]
```diff
@@ -1735,22 +1735,64 @@ static MethodHandle filterArgument(MethodHandle target, int pos, MethodHandle fi
...
+		
+		int callerLookupMode = callerLookup.lookupModes();
+		if (Lookup.MODULE != (Lookup.MODULE & callerLookupMode)) {
+			/*[MSG "K065W", "The access mode: 0x{0} of the caller lookup doesn't have the MODULE mode : 0x{1}"]*/
+			throw new IllegalAccessException(com.ibm.oti.util.Msg.getString("K065W", Integer.toHexString(callerLookupMode), Integer.toHexString(Lookup.MODULE))); //$NON-NLS-1$
+		}
+		
+		SecurityManager secmgr = System.getSecurityManager();
+		if (null != secmgr) {
+			secmgr.checkPermission(new ReflectPermission("suppressAccessChecks")); //$NON-NLS-1$
```
Comment (JasonFengJ9) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: This `RuntimePermission` can be defined within `com\ibm\oti\util\RuntimePermissions.java` as a static constant.

File: test/OpenJ9_Jsr_292_API/modules/mods.modulea/mods/modulea/package1/AnotherModuleExample.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
+/*******************************************************************************
+ * Copyright (c) 2017 IBM Corp. and others
```
Comment (JasonFengJ9) [inline test/OpenJ9_Jsr_292_API/modules/mods.modulea/mods/modulea/package1/AnotherModuleExample.java]: Pls use `2017, 2017` instead.

File: test/OpenJ9_Jsr_292_API/modules/mods.modulea/mods/modulea/package1/ModuleExample.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
+/*******************************************************************************
+ * Copyright (c) 2017 IBM Corp. and others
```
Comment (JasonFengJ9) [inline test/OpenJ9_Jsr_292_API/modules/mods.modulea/mods/modulea/package1/ModuleExample.java]: Same as comments above.

File: test/OpenJ9_Jsr_292_API/modules/mods.modulea/mods/modulea/package2/SameModuleExample.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
+/*******************************************************************************
+ * Copyright (c) 2017 IBM Corp. and others
```
Comment (JasonFengJ9) [inline test/OpenJ9_Jsr_292_API/modules/mods.modulea/mods/modulea/package2/SameModuleExample.java]: Same as comments above.

File: test/OpenJ9_Jsr_292_API/modules/mods.moduleb/mods/moduleb/package1/DifferentModuleExample1.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
+/*******************************************************************************
+ * Copyright (c) 2017 IBM Corp. and others
```
Comment (JasonFengJ9) [inline test/OpenJ9_Jsr_292_API/modules/mods.moduleb/mods/moduleb/package1/DifferentModuleExample1.java]: Same as comments above.

File: test/OpenJ9_Jsr_292_API/modules/mods.moduleb/module-info.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
+/*******************************************************************************
+ * Copyright (c) 2017 IBM Corp. and others
```
Comment (JasonFengJ9) [inline test/OpenJ9_Jsr_292_API/modules/mods.moduleb/module-info.java]: Same as comments above.

File: test/OpenJ9_Jsr_292_API/modules/mods.modulec/mods/modulec/package1/DifferentModuleExample2.java [hunk 10]
```diff
@@ -0,0 +1,28 @@
+/*******************************************************************************
+ * Copyright (c) 2017 IBM Corp. and others
```
Comment (JasonFengJ9) [inline test/OpenJ9_Jsr_292_API/modules/mods.modulec/mods/modulec/package1/DifferentModuleExample2.java]: Same as comments above.

File: test/OpenJ9_Jsr_292_API/modules/mods.modulec/module-info.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
+/*******************************************************************************
+ * Copyright (c) 2017 IBM Corp. and others
```
Comment (JasonFengJ9) [inline test/OpenJ9_Jsr_292_API/modules/mods.modulec/module-info.java]: Same as comments above.

File: test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_privateLookupIn.java [hunk 12]
```diff
@@ -0,0 +1,152 @@
+/*******************************************************************************
+ * Copyright (c) 2017 IBM Corp. and others
```
Comment (JasonFengJ9) [inline test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_privateLookupIn.java]: Same as comments above.

**Comment by JasonFengJ9** [comment]
Pls ensure new files using Unix delimiters.

File: jcl/src/java.base/share/classes/com/ibm/oti/util/ExternalMessages-MasterIndex.properties [hunk 13]
```diff
@@ -1239,6 +1239,11 @@ K065O=The parameter types of init doesn't match that of iterator: {0} != {1}
 K065P=The external parameter types of body: {0} doesn't match the external parameter list: {1}
 K065Q=The loop body must be non-null
 K065R=The requested lookup mode: 0x{0} is not one of the existing access modes: 0x{1}
+K065S=Both the requested class and the caller lookup must be non-null
```
Comment (JasonFengJ9) [inline jcl/src/java.base/share/classes/com/ibm/oti/util/ExternalMessages-MasterIndex.properties]: _must not be null_ sounds better.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 14]
```diff
@@ -1735,22 +1735,64 @@ static MethodHandle filterArgument(MethodHandle target, int pos, MethodHandle fi
...
+	/*[IF Sidecar19-SE-B175]*/
 	/**
-	 * Return a MethodHandles.Lookup object that is only able to access <code>private</code> members.
+	 * Return a MethodHandles.Lookup object with full capabilities including the access 
+	 * to the <code>private</code> members in the requested class
 	 * 
 	 * @param targetClass - the requested class containing private members
 	 * @param callerLookup - a Lookup object specific to the caller
 	 * @return a MethodHandles.Lookup object with private access to the requested class
+	 * @throws NullPointerException - if any of the arguments is null
```
Comment (JasonFengJ9) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Pls be specific about the argument, `targetClass` or `callerLookup`.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 15]
```diff
@@ -1735,22 +1735,64 @@ static MethodHandle filterArgument(MethodHandle target, int pos, MethodHandle fi
...
+		}
+		
+		Module targetClassModule = targetClass.getModule();
+		String targetClassPackageName = targetClass.getPackageName();
+		Module accessClassModule = callerLookup.accessClass.getModule();
+		
+		/* Check whether the named module containing the old lookup can read the module containing the target class.
+		 * Note: an unnamed module can read any module.
+		 */
+		if (accessClassModule.isNamed() && (!accessClassModule.canRead(targetClassModule))) {
```
Comment (JasonFengJ9) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: `accessClassModule.isNamed()` is redundant. `Module.canRead()` handles unnamed module case internally.

**Comment by AdamBrousseau** [comment]
Jenkins line endings check

File: test/OpenJ9_Jsr_292_API/playlist.xml (hunk 3 above)
Comment (ChengJin01) [reply test/OpenJ9_Jsr_292_API/playlist.xml]: Fixed.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 4 above)
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Agreed and fixed as suggested.

File: test/OpenJ9_Jsr_292_API/modules/mods.modulea/mods/modulea/package1/AnotherModuleExample.java (hunk 5 above)
Comment (ChengJin01) [reply test/OpenJ9_Jsr_292_API/modules/mods.modulea/mods/modulea/package1/AnotherModuleExample.java]: Updated.

File: test/OpenJ9_Jsr_292_API/modules/mods.modulea/mods/modulea/package1/ModuleExample.java (hunk 6 above)
Comment (ChengJin01) [reply test/OpenJ9_Jsr_292_API/modules/mods.modulea/mods/modulea/package1/ModuleExample.java]: Updated.

File: test/OpenJ9_Jsr_292_API/modules/mods.moduleb/mods/moduleb/package1/DifferentModuleExample1.java (hunk 8 above)
Comment (ChengJin01) [reply test/OpenJ9_Jsr_292_API/modules/mods.moduleb/mods/moduleb/package1/DifferentModuleExample1.java]: Updated.

File: test/OpenJ9_Jsr_292_API/modules/mods.modulea/mods/modulea/package2/SameModuleExample.java (hunk 7 above)
Comment (ChengJin01) [reply test/OpenJ9_Jsr_292_API/modules/mods.modulea/mods/modulea/package2/SameModuleExample.java]: Updated.

File: test/OpenJ9_Jsr_292_API/modules/mods.moduleb/module-info.java (hunk 9 above)
Comment (ChengJin01) [reply test/OpenJ9_Jsr_292_API/modules/mods.moduleb/module-info.java]: Updated.

File: test/OpenJ9_Jsr_292_API/modules/mods.modulec/mods/modulec/package1/DifferentModuleExample2.java (hunk 10 above)
Comment (ChengJin01) [reply test/OpenJ9_Jsr_292_API/modules/mods.modulec/mods/modulec/package1/DifferentModuleExample2.java]: Updated.

File: test/OpenJ9_Jsr_292_API/modules/mods.modulec/module-info.java (hunk 11 above)
Comment (ChengJin01) [reply test/OpenJ9_Jsr_292_API/modules/mods.modulec/module-info.java]: Updated.

File: jcl/src/java.base/share/classes/com/ibm/oti/util/ExternalMessages-MasterIndex.properties (hunk 13 above)
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/com/ibm/oti/util/ExternalMessages-MasterIndex.properties]: Updated.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 14 above)
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Updated.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 15 above)
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Agreed and fixed.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 16]
```diff
@@ -1735,22 +1735,64 @@ static MethodHandle filterArgument(MethodHandle target, int pos, MethodHandle fi
...
+		
+		int callerLookupMode = callerLookup.lookupModes();
+		if (Lookup.MODULE != (Lookup.MODULE & callerLookupMode)) {
+			/*[MSG "K065W", "The access mode: 0x{0} of the caller lookup doesn't have the MODULE mode : 0x{1}"]*/
+			throw new IllegalAccessException(com.ibm.oti.util.Msg.getString("K065W", Integer.toHexString(callerLookupMode), Integer.toHexString(Lookup.MODULE))); //$NON-NLS-1$
+		}
+		
+		SecurityManager secmgr = System.getSecurityManager();
+		if (null != secmgr) {
+			secmgr.checkPermission(com.ibm.oti.util.ReflectPermissions.permissionSuppressAccessChecks);
```
Comment (JasonFengJ9) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: There is another occurrence of `new ReflectPermission("suppressAccessChecks")` at method `reflectAs`, pls change it to `ReflectPermissions.permissionSuppressAccessChecks` as well.

**Comment by JasonFengJ9** [comment]
lgtm

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 16 above)
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Updated.

**Comment by DanHeidinga** [comment]
@ChengJin01 Can you fix the line endings in the following file:
```
test/OpenJ9_Jsr_292_API/modules/mods.modulea/module-info.java
test/OpenJ9_Jsr_292_API/modules/mods.moduleb/module-info.java
test/OpenJ9_Jsr_292_API/modules/mods.modulec/module-info.java
```
so the Line Endings Check will pass?

Jenkins test sanity

**Comment by DanHeidinga** [comment]
@ChengJin01 This is ready to merge once the line endings are fixed.

**Comment by ChengJin01** [comment]
Already updated the ending in module-info.java

**Comment by ChengJin01** [comment]
It seems the line ending check still failed after fixing the ending in module-info.java

**Comment by ChengJin01** [comment]
The issue with line endings was fixed.

**Pull request merged**